posts/2021/09/array-libraries-interoperability.md

```diff
@@ -0,0 +1,578 @@
+and
+[NEP 37](https://numpy.org/neps/nep-0037-array-module.html) endeavouring to
+address some of the issues and shortcomings with NEP 18. Note that these NEPs
+were actually never accepted or implemented.
+
+**For the sake of brevity in this blog,
+we'll limit our focus to [NEP 31 or `uarray`](https://uarray.org/en/latest/)
+and [NEP 47 or Array API (`__array_namespace__`)](https://numpy.org/neps/nep-0047-array-api-standard.html). These are some of the most recent protocols
+with a goal to ameliorate interoperability shortfalls.**
+
+
+## Array API or NEP 47
+
+Before I start describing Python Array API in the context of this blog,
+I urge you to read:
+
+1. [Announcing the consortium](https://data-apis.org/blog/announcing_the_consortium/)
+2. [First release of the Array API Standard](https://data-apis.org/blog/array_api_standard_release/) 
+
+These two official blogs from the [Consortium for Python Data API Standards](https://data-apis.org/) describe the API, giving a high level overview of its existence.
+
+Let's see how the use of this Array API might look in practice.
+
+```python
+import torch as xp
+# or
+import numpy.array_api as xp
+
+a = xp.arange(3)
+b = xp.ones(3)
+
+c = a + b
+``` 
+
+Technically, the only changes involved for a NumPy end user to port their
+code in PyTorch would be to update the import statement and refactor
+`np.` -> `xp.`. Here `xp` represents any array provider library
+compliant with the Array API. Doing something like this is extremely easy as
+compared to some other array interoperability protocols taking a much more
+convoluted approach.
+
+A couple of the concrete use cases mentioned in the [Array API Spec](https://data-apis.org/array-api/latest/use_cases.html#concrete-use-cases) are:
+
+* Use case 1: add hardware accelerator and distributed support to SciPy
+* Use case 2: simplify einops by removing the backend system
+
+With the introduction and compliance of Array API in all the major
+array/tensor libraries, consumer libraries will be able to support more
+than one "array provider" and become truly interoperable.
+
+Since we started by talking about consumer libraries like SciPy, let's
+continue with the same example. We've built a [demo](https://anirudhdagar.ml/array-api-demo/intro.html) around Array API showcasing the use of PyTorch Tensors with
+SciPy.
+
+
+<p align="center">
+	<a href="https://anirudhdagar.ml/array-api-demo/intro.html">
+	<img alt="Array API Prototype" src="/images/2021/09/array_api_demo_screenshot.png">
+    <i>Array API Demo</i>
+    </a>
+</p>
+
+The [Array API Demo](https://anirudhdagar.ml/array-api-demo/intro.html) walks
+you through the details and processes involved to make an array consumer library
+like SciPy more interoperable with array provider libraries. The demo is built
+keeping two different perspectives in mind, one that of an end user and the other
+of an open-source developer/maintainer looking to incorporate and use Array API
```

Comment:
```suggestion
to keep two different perspectives in mind: an end user, and
an open-source developer/maintainer looking to incorporate use of the Array API
```

Reply:
With this line, I'm trying to say: "With these two different perspectives in mind, we've built the demo."
It's not the other way round. Basically, the demo is built to show two mindsets, one that of the user and the other of the maintainer.

I'm not super sure if your suggestion means exactly the same.

Reply:
Ahh I shouldn't of replaced "keeping" to "to keep".
```suggestion
keeping two different perspectives in mind: an end user, and
an open-source developer/maintainer looking to incorporate use of the Array API
```
Does this look better? I *think* I understood what you mean, this was meant to just help readability :sweat_smile:

Reply:
Yeah this looks good. Thanks for the quick update.